eventlet/websocket.py

```diff
@@ -528,25 +528,31 @@ def _get_permessage_deflate_enc(self):
 
         def _make():
             return zlib.compressobj(zlib.Z_DEFAULT_COMPRESSION, zlib.DEFLATED,
-                                    -options.get("server_max_window_bits", zlib.MAX_WBITS))
+                                    -options.get("client_max_window_bits" if self.client
+                                                 else "server_max_window_bits",
+                                                 zlib.MAX_WBITS))
 
-        if options.get("server_no_context_takeover"):
+        if options.get("client_no_context_takeover" if self.client
+                       else "server_no_context_takeover"):
             # This option means we have to make a new one every time
             return _make()
         else:
             if self._deflate_enc is None:
                 self._deflate_enc = _make()
             return self._deflate_enc
 
-    def _get_permessage_deflate_dec(self):
+    def _get_permessage_deflate_dec(self, rsv1):
         options = self.extensions.get("permessage-deflate")
-        if options is None:
+        if options is None or not rsv1:
             return None
 
         def _make():
-            return zlib.decompressobj(-options.get("client_max_window_bits", zlib.MAX_WBITS))
+            return zlib.decompressobj(-options.get("server_max_window_bits" if self.client
```

Comment:
It says `server_... if self.client`. Is that correct?

Reply:
Yes, this is correct. During decompression, we want to be using the same settings as our counterparty. So if the websocket is acting as it client, it should use the negotiated server settings during decompression and vice versa.

Technically, during decompression, either party could be using a larger LZ77 sliding window and always allow context takeover, but limiting it to the agreed setting of the other party helps save on memory needed for the decompression.

Reply:
Thanks for explain.